### PR TITLE
Refactoring: LDAP Connection Management

### PR DESCRIPTION
**Pull request by tomaskrizek**
Design doc: http://www.freeipa.org/page/V4/LDAP_Connection_Management_Refactoring

Many changes in current PR are only temporary for easier refactoring, but feedback is welcome.

### CODE REVIEW

File: ipaserver/install/dsinstance.py [hunk 1]
```diff
@@ -265,12 +261,14 @@ def __common_setup(self, enable_ssl=False):
         self.step("creating directory server instance", self.__create_instance)
         self.step("updating configuration in dse.ldif", self.__update_dse_ldif)
         self.step("restarting directory server", self.__restart_instance)
+        self.step("enabling ldapi", self.__enable_ldapi)
```
Comment (tbordaz) [inline ipaserver/install/dsinstance.py]: Hello,

Isn't it possible to enable ldapi+autobind before the 'line 263' restart ?
I would save a DS restart

**Comment by mbasti-rh** [comment]
Please fix PEP8
```
./ipapython/ipaldap.py:766:49: E231 missing whitespace after ','
./ipapython/ipaldap.py:822:9: E265 block comment should start with '# '
./ipaserver/install/bindinstance.py:231:1: E302 expected 2 blank lines, found 1
./ipaserver/install/bindinstance.py:231:80: E501 line too long (82 > 79 characters)
./ipaserver/install/krbinstance.py:170:80: E501 line too long (89 > 79 characters)
./ipaserver/install/ldapupdate.py:59:80: E501 line too long (85 > 79 characters)
./ipaserver/install/replication.py:218:80: E501 line too long (82 > 79 characters)
```

File: ipapython/ipaldap.py [hunk 2]
```diff
@@ -1639,48 +1636,25 @@ def __init__(self, host='', port=389, cacert=None, debug=None, ldapi=False,
     def __str__(self):
         return self.host + ":" + str(self.port)
 
-    def __wait_for_connection(self, timeout):
```
Comment (mbasti-rh) [inline ipapython/ipaldap.py]: LGTM
https://fedorahosted.org/freeipa/ticket/2175

Upgrade is using ldapi now.

However I'm curious if there is a time delay between DS is listening on ldapi socket and when DS is listening on 389 port. But in our case we plan to use just ldapi, so it should be ok.

File: ipapython/ipaldap.py [hunk 3]
```diff
@@ -704,7 +704,7 @@ class LDAPClient(object):
     time_limit = -1.0   # unlimited
     size_limit = 0      # unlimited
 
-    def __init__(self, ldap_uri, start_tls=False, force_schema_updates=False,
+    def __init__(self, ldap_uri, start_tls=False, force_schema_updates=True,
```
Comment (mbasti-rh) [inline ipapython/ipaldap.py]: Please add reason for this change into commit message

File: ipapython/ipaldap.py [hunk 4]
```diff
@@ -704,8 +704,44 @@ class LDAPClient(object):
...
+        """
+        if self.cacert is not None:
+            return 'ldaps'
+        elif self.ldapi:
+            return 'ldapi'
+        else:
+            return 'ldap'
+
+    def __init__(self, host='', start_tls=False,
+                 force_schema_updates=True, no_schema=False,
```
Comment (mbasti-rh) [inline ipapython/ipaldap.py]: I think this comment should be in previous patch

File: ipaserver/install/dsinstance.py (hunk 1 above)
Comment (mbasti-rh) [inline ipaserver/install/dsinstance.py]: Agree, I'd like to avoid additional DS restart as much as possible

File: ipaserver/install/service.py [hunk 5]
```diff
@@ -157,7 +158,8 @@ def ldap_disconnect(self):
         LDAPConnectionManager.close_connection()
         self.admin_conn = None
 
-    def _ldap_mod(self, ldif, sub_dict=None, raise_on_err=True):
+    def _ldap_mod(self, ldif, sub_dict=None, raise_on_err=True,
```
Comment (mbasti-rh) [inline ipaserver/install/service.py]: I don't like to ahve so much options, can we split this into separate method _simple_ldap_mod(), _nonldapi_, or something

File: ipapython/ipaldap.py [hunk 6]
```diff
@@ -1656,3 +1656,27 @@ def entry_exists(self, dn):
             return False
         else:
             return True
+
+
+class LDAPConnectionManager(object):
```
Comment (mbasti-rh) [inline ipapython/ipaldap.py]: Because this is related only to ldapi, I would rename it to LDAPI connection manager

**Comment by mbasti-rh** [comment]
I did some inline comments, I'm not fully satisfied with implementation of the connection manager, I'll think about it tomorrow.

File: ipaserver/install/dsinstance.py (hunk 1 above)
Comment (tomaskrizek) [inline ipaserver/install/dsinstance.py]: I've tried to do that, but I haven't been able to connect to LDAP before the first restart happens.

Comment (tomaskrizek) [inline ipaserver/install/dsinstance.py]: You're right, it is possible to connect to ldap right after the instance is created. I didn't notice the service is stopped before updating the dse. Thanks, fixed in next synchronization.

File: ipapython/ipaldap.py (hunk 3 above)
Comment (tomaskrizek) [inline ipapython/ipaldap.py]: Fixed in next sync.

**Comment by tomaskrizek** [comment]
I made some changes and removed the connection manager, as discussed with sub-team. The refactoring is not finished, but these changes can be pushed to master if it doesn't break any working tests. Jobs are currently pending in jenkins.

**Comment by tomaskrizek** [comment]
Please review the code. I fixed a problem that was identified with the previous set of integration tests, but I'm still waiting for results of this build. Nevertheless, no more changes should be introduced in this PR, only potential fixes of issues.

**Comment by rcritten** [comment]
Removing the connection timeout makes me a bit edgy. It is quite unclear why that's there, perhaps python-ldap doesn't provide a timeout and things hung forever, I'm not sure. I'd double-check that there is a timeout if the remote server either isn't there (probably raises a connection error), or is there but doesn't respond.

File: ipaserver/install/server/replicainstall.py [hunk 7]
```diff
@@ -1030,7 +1031,13 @@ def promote_check(installer):
         except ipaclient.ntpconf.NTPConfigurationError:
             pass
 
-    api.bootstrap(in_server=True, context='installer')
+    env = Env()
+    env._bootstrap(context='installer', log=None)
+    env._finalize_core(**dict(constants.DEFAULT_CONFIG))
+
+    # pylint: disable=no-member
```
Comment (mbasti-rh) [inline ipaserver/install/server/replicainstall.py]: NACK: you disabled pylint globally until end of file, enable it later or put it to the right line

File: ipaserver/install/dsinstance.py [hunk 8]
```diff
@@ -670,7 +670,7 @@ def apply_updates(self):
             # very fatal errors only will raise exception
             raise RuntimeError("Update failed: %s" % e)
         installutils.store_version()
-
+        self.ldap_disconnect()
```
Comment (mbasti-rh) [inline ipaserver/install/dsinstance.py]: Why it is located here, I don't see DS restart in this function, IMO this disconnect should be as close to restart as possible (i.e. in the same function as restart)

**Comment by mbasti-rh** [comment]
@rcritten Tomas removed timelimit that was used for repeated connections, it is not used for preventing hangs. (If we talk about the same commit 'ldap refactoring: remove wait/timeout during binds') We added this functionality to DS restart, restart will block code until DS is not ready on LDAPI port.

@tomaskrizek I put two inline comments there, otherwise changes make sense. I'll wait for tests results

File: install/restart_scripts/restart_dirsrv [hunk 9]
```diff
@@ -38,6 +38,8 @@ def _main():
 
     syslog.syslog(syslog.LOG_NOTICE, "certmonger restarted dirsrv instance '%s'" % instance)
 
+    api.Backend.ldap2.close()
```
Comment (jcholast) [inline install/restart_scripts/restart_dirsrv]: On `ldap2`, `.disconnect()` should be called rather than `.close()`.

File: ipapython/ipaldap.py [hunk 10]
```diff
@@ -735,6 +787,46 @@ def __init__(self, ldap_uri, start_tls=False, force_schema_updates=False,
...
+        with self.error_handler():
+            if demand_cert:
+                ldap.set_option(ldap.OPT_X_TLS_REQUIRE_CERT, True)
+                self.conn.set_option(ldap.OPT_X_TLS_DEMAND, True)
+
+            if sasl_nocanon:
+                self.conn.set_option(ldap.OPT_X_SASL_NOCANON, ldap.OPT_ON)
+
+            if start_tls:
+                self.conn.start_tls_s()
```
Comment (jcholast) [inline ipapython/ipaldap.py]: This block should either be moved to `_connect()`, or the rest of `_connect()` should be moved here.

**Comment by tomaskrizek** [comment]
In an offline discussion we decided not to push temporary changes to master. Here's the final code for review.

File: ipaserver/plugins/ldap2.py [hunk 11]
```diff
@@ -75,8 +77,8 @@ def __init__(self, api, ldap_uri=None):
         LDAPClient.__init__(self, ldap_uri,
                             force_schema_updates=force_schema_updates)
 
-        self.__time_limit = None
-        self.__size_limit = None
+        self.__time_limit = float(LDAPClient.time_limit)
+        self.__size_limit = float(LDAPClient.size_limit)
```
Comment (jcholast) [inline ipaserver/plugins/ldap2.py]: Size limit is an `int`, not `float`.

File: ipapython/ipaldap.py [hunk 12]
```diff
@@ -723,18 +721,121 @@ def __init__(self, ldap_uri, start_tls=False, force_schema_updates=False,
...
+
+    @staticmethod
+    def IPAdmin_init(host='', port=389, cacert=None, debug=None,
```
Comment (jcholast) [inline ipapython/ipaldap.py]: This should be a function called `IPAdmin` rather than a static method of `LDAPClient`, that way you don't have to modify any of the existing code which calls `IPAdmin`.

File: pylintrc [hunk 13]
```diff
@@ -27,6 +27,7 @@ disable=
     too-many-boolean-expressions,
     too-many-branches,
     too-many-instance-attributes,
+    too-many-function-args,
```
Comment (jcholast) [inline pylintrc]: Why is this necessary?

File: ipapython/ipaldap.py [hunk 14]
```diff
@@ -1085,7 +1085,7 @@ def simple_bind(self, bind_dn=DN(('cn', 'directory manager')),
             self.conn.simple_bind_s(
                 bind_dn, bind_password, server_controls, client_controls)
 
-    def external_bind(self, user_name, server_controls=None,
+    def external_bind(self, user_name=None, server_controls=None,
```
Comment (jcholast) [inline ipapython/ipaldap.py]: Remove the `user_name` argument altogether, as it always has to be set to the effective username.

File: ipapython/ipaldap.py [hunk 15]
```diff
@@ -1068,7 +1069,8 @@ def _connect(self):
 
         return conn
 
-    def simple_bind(self, bind_dn, bind_password, server_controls=None,
+    def simple_bind(self, bind_dn=DN(('cn', 'directory manager')),
+                    bind_password='', server_controls=None,
```
Comment (jcholast) [inline ipapython/ipaldap.py]: I would really prefer if both `bind_dn` and `bind_password` had to be always specified explicitly.

File: ipapython/ipaldap.py [hunk 16]
```diff
@@ -720,18 +721,121 @@ def __init__(self, ldap_uri, start_tls=False, force_schema_updates=False,
...
+            cacert=cacert,
+            sasl_nocanon=sasl_nocanon,
+            demand_cert=demand_cert)
+
+        ldap_client.port = port
+        ldap_client.host = host
+        ldap_client.cacert = cacert
+        ldap_client.ldapi = ldapi
+        ldap_client.realm = realm
+        ldap_client.suffixes = {}
```
Comment (jcholast) [inline ipapython/ipaldap.py]: Are all of these really necessary? `git grep` shows me that `.cacert`, `.realm` and `.suffixes` are not used anywhere and `.ldapi` is used only in `ipa-httpd-kdcproxy`. That leaves us with `.host` and `.port`, which should preferrably be transformed into properties of `LDAPClient` rather than set here.

File: ipaserver/install/dsinstance.py [hunk 17]
```diff
@@ -615,14 +615,19 @@ def __add_default_schemas(self):
             # Does not apply with newer DS releases
             pass
 
+    def start(self, *args, **kwargs):
+        super(DsInstance, self).start(*args, **kwargs)
+        api.Backend.ldap2.connect(size_limit=LDAPClient.size_limit,
+                                  time_limit=LDAPClient.time_limit)
```
Comment (jcholast) [inline ipaserver/install/dsinstance.py]: These are the default limits in `ldap2`. Why do you set them again here?

File: ipaserver/plugins/ldap2.py [hunk 18]
```diff
@@ -102,7 +106,9 @@ def size_limit(self):
 
     @size_limit.setter
     def size_limit(self, val):
-        self.__size_limit = int(val)
+        if val is not None:
+            val = int(val)
+        self.__size_limit = val
 
     @size_limit.deleter
```
Comment (jcholast) [inline ipaserver/plugins/ldap2.py]: You should also update the deleters.

File: ipaserver/plugins/ldap2.py [hunk 19]
```diff
@@ -150,7 +150,7 @@ def create_connection(
         Extends backend.Connectible.create_connection.
         """
         if bind_dn is None:
-            bind_dn = DN()
+            bind_dn = DN(('cn', 'directory manager'))
```
Comment (jcholast) [inline ipaserver/plugins/ldap2.py]: Actually, if `bind_dn` is None, simple bind should not be attempted.

File: ipaserver/install/replication.py [hunk 20]
```diff
@@ -505,7 +505,7 @@ def setup_changelog(self, conn):
     def setup_chaining_backend(self, conn):
         chaindn = DN(('cn', 'chaining database'), ('cn', 'plugins'), ('cn', 'config'))
         benamebase = "chaindb"
-        urls = [self.to_ldap_url(conn)]
+        urls = [self.to_ldap_url()]
```
Comment (jcholast) [inline ipaserver/install/replication.py]: I would rather throw away `self.to_ldap_url()` and use `conn.ldap_uri` here.

File: ipapython/ipaldap.py (hunk 12 above)
Comment (tomaskrizek) [inline ipapython/ipaldap.py]: Well this code was already modified, twice. I do not see the value in changing this yet again.

File: pylintrc (hunk 13 above)
Comment (tomaskrizek) [inline pylintrc]: `IPAdmin_init` has too many arguments. This check didn't trigger previously when it was in `IPAdmin.__init__`

**Comment by jcholast** [comment]
In addition to my inline comments:
* use component name ("ipaldap", "ldap2", "install", ...) rather than "lda refactoring" as a prefix in commit subjects,
* please move "ldap refactoring: change default time/size limit in ldap2" before "ldap refactoring: conn management in dsinstance" and squash it with "ldap refactoring: restore previous time/size limit in backend",
* squash "ldap refactoring: add restart_dirsrv to installutils" and "ldap refactoring: use restart_dirsrv in installers",
* maybe squash all of the "ldap refactoring: conn management in ipa-...", as it is a single change accross multiple scripts.

File: ipapython/ipaldap.py (hunk 15 above)
Comment (tomaskrizek) [inline ipapython/ipaldap.py]: I don't think I've encountered a place in the code where `bind_dn` would be anything other than `cn=directory manager`. I think having sane defaults is a good approach in that case.

File: ipaserver/plugins/ldap2.py (hunk 11 above)
Comment (tomaskrizek) [inline ipaserver/plugins/ldap2.py]: Ok.

File: ipaserver/install/dsinstance.py (hunk 17 above)
Comment (tomaskrizek) [inline ipaserver/install/dsinstance.py]: That's a temporary change that is refactored in a later commit, the final code calls `api.Backend.ldap2.connect()`.

File: ipaserver/plugins/ldap2.py (hunk 19 above)
Comment (tomaskrizek) [inline ipaserver/plugins/ldap2.py]: Ok.

File: ipaserver/install/dsinstance.py [hunk 21]
```diff
@@ -257,14 +257,15 @@ def __common_setup(self, enable_ssl=False):
 
         self.step("creating directory server user", create_ds_user)
         self.step("creating directory server instance", self.__create_instance)
+        self.step("enabling ldapi", self.__enable_ldapi)
+        self.step("configure autobind for root", self.__root_autobind)
```
Comment (tomaskrizek) [inline ipaserver/install/dsinstance.py]: That would be a nice change, as well as removing `_ldap_mod()` completely, but it's out of scope for this refactoring.

File: pylintrc (hunk 13 above)
Comment (tomaskrizek) [inline pylintrc]: No longer necessary.

File: ipapython/ipaldap.py (hunk 14 above)
Comment (tomaskrizek) [inline ipapython/ipaldap.py]: Ok.

File: ipapython/ipaldap.py [hunk 22]
```diff
@@ -720,18 +721,121 @@ def __init__(self, ldap_uri, start_tls=False, force_schema_updates=False,
             If true, attributes are decoded to Python types according to their
             syntax.
         """
+        if debug and debug.lower() == "on":
+            ldap.set_option(ldap.OPT_DEBUG_LEVEL, 255)
+        if cacert is not None:
+            ldap.set_option(ldap.OPT_X_TLS_CACERTFILE, cacert)
```
Comment (tomaskrizek) [inline ipapython/ipaldap.py]: Fixed.

File: ipapython/ipaldap.py (hunk 16 above)
Comment (tomaskrizek) [inline ipapython/ipaldap.py]: Moved `host` and `port` to `LDAPClient`. Removed the rest.

File: ipaserver/install/replication.py (hunk 20 above)
Comment (tomaskrizek) [inline ipaserver/install/replication.py]: Ok.

**Comment by tomaskrizek** [comment]
I've made some modifications based on the feedback. Since I've changed a lot of code, I've once again tested some use cases manually and fixed all the bugs I was able to find. Jenkins might stumble upon some problems, but hopefully we can merge this ASAP. I'm on PTO on Friday so if we want to merge this week, I can make some last modifications tomorrow.

File: ipaserver/install/server/install.py [hunk 23]
```diff
@@ -752,6 +752,7 @@ def install(installer):
         ntpinstance.ntp_ldap_enable(host_name, ds.suffix, realm_name)
 
     else:
+        api.Backend.ldap2.connect()
```
Comment (tomaskrizek) [inline ipaserver/install/server/install.py]: Only change in last sync.

**Comment by tomaskrizek** [comment]
I've also updated the design document and described the correct usage of LDAP connections after these changes:

http://www.freeipa.org/page/V4/LDAP_Connection_Management_Refactoring

**Comment by mbasti-rh** [comment]
I don't like `_missing` value in ldap2 class for time and sizelimit

IIUC then `_missing` means that by default the search is unlimited. Why we just dont create constants for unlimited search and use it, it is more clear than combinations of three types `_missing`, Int and None

```
class LDAPClient(object):

SIZE_NOLIMIT = 0
TIME_NOLIMIT = -1.0


class ldap2(CrudBackend, LDAPClient)

   def create_connection(..., time_limit=TIME_NOLIMIT, size_limit=SIZE_NOLIMIT):
   ...
```

**Comment by mbasti-rh** [comment]
LGTM except my previous comment, lets wait for test results

**Comment by jcholast** [comment]
@mbasti: YDUC, that would break `ipa config`-defined limits.

**Comment by mbasti-rh** [comment]
@jcholast how? With this patchset you have to set `None` to time_limit/size_limit to apply `ipa config` limits because `_missing`  means "unlimited"

```
        if time_limit is not _missing:
            self.time_limit = time_limit
```

So with _missing default value is used, what is 0 (unlimited)
```
self.__time_limit = float(LDAPClient.time_limit)
```

**Comment by jcholast** [comment]
@mbasti: Because if the argument default values were constants as you suggested, they would take precedence over `ipa config` limits.

**Comment by pvoborni** [comment]
We have a lot of refactoring patches which needs to be serialized and pushed and we don't have much time for it.

If the remaining issue doesn't break everything else and can be fixed by small patch later then please push this and allow other efforts to rebase on it.

**Comment by mbasti-rh** [comment]
NACK

Multiple tests are failing with:
```
TypeError: 'int' object has no attribute '__getitem__'

self = <ipatests.test_integration.test_backup_and_restore.TestBackupAndRestore object at 0x7f61d7dade10>

    def test_full_backup_and_restore(self):
        """backup, uninstall, restore"""
>       with restore_checker(self.master):

test_integration/test_backup_and_restore.py:146: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib64/python2.7/contextlib.py:17: in __enter__
    return self.gen.next()
test_integration/test_backup_and_restore.py:114: in restore_checker
    results.append(check(host))
test_integration/test_backup_and_restore.py:53: in check_admin_in_ldap
    ldap = host.ldap_connect()
test_integration/host.py:47: in ldap_connect
    ldap_uri = ipaldap.LDAPClient(self.external_hostname)
../ipapython/ipaldap.py:747: in __init__
    self._conn = self._connect()
../ipapython/ipaldap.py:1120: in _connect
    conn.start_tls_s()
/usr/lib64/python2.7/contextlib.py:35: in __exit__
    self.gen.throw(type, value, traceback)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <ipapython.ipaldap.LDAPClient object at 0x7f61d7d8d8d0>, arg_desc = None

    @contextlib.contextmanager
    def error_handler(self, arg_desc=None):
        """Context manager that handles LDAPErrors
            """
        try:
            try:
                yield
            except ldap.TIMEOUT:
                raise errors.DatabaseTimeout()
            except ldap.LDAPError as e:
>               desc = e.args[0]['desc'].strip()
E               TypeError: 'int' object has no attribute '__getitem__'

../ipapython/ipaldap.py:995: TypeError
```

and

replica installations fails in some cases with

```
<DEBUG>:   [5/7]: enable GSSAPI for replication
<DEBUG>:   [error] OBJECT_CLASS_VIOLATION: {'desc': 'Object class violation'}
<DEBUG>: ipa.ipapython.install.cli.install_tool(Replica): ERROR    {'desc': 'Object class violation'}
<DEBUG>: ipa.ipapython.install.cli.install_tool(Replica): ERROR    The ipa-replica-install command failed. See /var/log/ipareplica-install.log for more information
```


None of above is happening in master branch tests.

**Comment by mbasti-rh** [comment]
INFO related to LDAPError:

```
-> desc = e.args[0]['desc'].strip()
(Pdb) print e
LDAPError(2, 'No such file or directory')
(Pdb) print self.ldap_uri
'master.ipa.test'
```
Probaly we should open ticket and improve LDAP error handling to cover this issue

error is in test_integration/host.py:ldap_connect
```
        ldap_uri = ipaldap.LDAPClient(self.external_hostname)
        ldap = ipaldap.LDAPClient(ldap_uri)  <-------- you created ldap_uri as LDAPClient object
```

**Comment by tomaskrizek** [comment]
- Fixed the `LDAPClient` typo; affected integration tests now pass
- Not able to reproduce the `[5/7]: enable GSSAPI for replication` replica installation error nor locally (tried twice), nor in jenkins
- Improved doc text for the `_missing` idiom in `ldap2.create_connection()`
- Ran integration tests again. The only difference from master are 4 failing tests in caless domlelv1: `test_caless.TestReplicaInstall`. I'm not able to reproduce these locally, but they seem to occur persistently in Jenkins. I wonder if they might be caused by some configuration or reusing VMs.

**Comment by mbasti-rh** [comment]
It was decided to push this and unblock other developer, minor issues will be addressed separately.

**Comment by tomaskrizek** [comment]
Updated commit messages with link to a ticket.

**Comment by mbasti-rh** [comment]
Fixed upstream
master:
https://fedorahosted.org/freeipa/changeset/5760b7e983da6bda8f5383d9079551e4acb4c2da
https://fedorahosted.org/freeipa/changeset/de58a5c60596de8b45c8016c3318bac78305477a
https://fedorahosted.org/freeipa/changeset/60e38ecc7ff6b983f4f3af0a66c08eb3a3fda22d
https://fedorahosted.org/freeipa/changeset/4f1a6a177666c475156f496d3f7719b37e66a7b0
https://fedorahosted.org/freeipa/changeset/5b81dbfda1e4f0799d4ce87e9987a896af3ff299
https://fedorahosted.org/freeipa/changeset/9340a1417acf120fed3e9ffbe9d658d3456743a1
https://fedorahosted.org/freeipa/changeset/24baccbd6ac8a19ba52619a3cc59366220c4ca1f
https://fedorahosted.org/freeipa/changeset/9fca820b6bc2144cd827bddba69cb53f8ba3f42a
https://fedorahosted.org/freeipa/changeset/7a1c0db989cf59a778676635e160f73ebc610694
https://fedorahosted.org/freeipa/changeset/e2780b2106a6e6bab0cb3f3d3ec06482cde9d374
https://fedorahosted.org/freeipa/changeset/8934d03b3b5bbf02e9e20a1644ef31d27fa0f483
https://fedorahosted.org/freeipa/changeset/e8aa2627c7a3dcb0b0745e656ea58ccbbccd38fb
https://fedorahosted.org/freeipa/changeset/e05bdeb6cf4505ef84e485b95b37aabba625160b
https://fedorahosted.org/freeipa/changeset/a77469f5984b12e201a3d349efad1ca2925ee5af
https://fedorahosted.org/freeipa/changeset/df86efdc69271cca0774868ab85b5be7df529136
https://fedorahosted.org/freeipa/changeset/49ff159a5f0cfd2f9d037ad00e75d8ac5bfba585
https://fedorahosted.org/freeipa/changeset/c51b04fae77149a09e921495c5b3c9802d199076
https://fedorahosted.org/freeipa/changeset/03d113cdd7c5f943d8937eb4fec1086bfe47e909
https://fedorahosted.org/freeipa/changeset/1240262a0b01ff8408c06058d6d4d61fc5cde548
https://fedorahosted.org/freeipa/changeset/36d95472d983ff342a43a5df36d932b9de8c32ac
https://fedorahosted.org/freeipa/changeset/922062eb559d1bb82a9d787763aacb31c0cf9b8d
https://fedorahosted.org/freeipa/changeset/7d028992ea2c2bf6acabe79f101621bdebbf9dbc
https://fedorahosted.org/freeipa/changeset/a9585ec563d1e54c3cd7de14789457f72cd00843
https://fedorahosted.org/freeipa/changeset/41098e3f7bb517f7445ed34d555bc3fb2083c6ce
